File: perl/words/t/words_test.pl
#!/usr/bin/perl
use strict;
use warnings;

require './lib/words.pl';

use Test::More tests => 18;

is(reverse_string("hello"), "olleh", "hello");
is(reverse_string("a"), "a", "a");
is(reverse_string(""), "", "");

is(is_palindrome("race car"), 1, "race car");
is(is_palindrome("level"), 1, "level");
is(is_palindrome("hello"), 0, "hello");
is(is_palindrome("a"), 1, "a");
is(is_palindrome(""), 1, "");

is(kmp_search("test", "this is a test"), 1, "is substring 1");
is(kmp_search("not", "this is a test"), 0, "is substring 2");
is(kmp_search("", "any string"), 1, "is substring 3");
is(kmp_search("abc", "abc"), 1, "is substring 4");
is(kmp_search("abc", "ab"), 0, "is substring 5");

is(lcs("AGGTAB", "GXTXAYB"), 4, "lcs 1");
is(lcs("ABC", "ABC"), 3, "lcs 2");
is(lcs("ABC", "DEF"), 0, "lcs 3");
is(lcs("", "ABC"), 0, "lcs 4");
is(lcs("ABC", ""), 0, "lcs 5");

File: _experimental/_c/words/test/words_test.c
#include <check.h>
#include <stdlib.h>
#include "../include/words.h"

START_TEST(test_reverse_string) {
    char* reversed_string = reverse_string("hello");
    ck_assert_str_eq(reversed_string, "olleh");
    free(reversed_string);

    reversed_string = reverse_string("a");
    ck_assert_str_eq(reversed_string, "a");
    free(reversed_string);

    reversed_string = reverse_string("");
    ck_assert_str_eq(reversed_string, "");
    free(reversed_string);
}

START_TEST(test_is_palindrome) {
    ck_assert(is_palindrome("race car") == true);
    ck_assert(is_palindrome("level") == true);
    ck_assert(is_palindrome("hallo") == false);
    ck_assert(is_palindrome("a") == true);
    ck_assert(is_palindrome("") == true);
}

START_TEST(test_is_substring) {
    ck_assert(kmp_search("test", "this is a test") == true);
    ck_assert(kmp_search("not", "this is a test") == false);
    ck_assert(kmp_search("", "any string") == true);
    ck_assert(kmp_search("abc", "abc") == true);
    ck_assert(kmp_search("abc", "ab") == false);
}

START_TEST(test_lcs) {
    ck_assert_int_eq(lcs("AGGTAB", "GXTXAYB"), 4);
    ck_assert_int_eq(lcs("ABC", "ABC"), 3);
    ck_assert_int_eq(lcs("ABC", "DEF"), 0);
    ck_assert_int_eq(lcs("", "ABC"), 0);
    ck_assert_int_eq(lcs("ABC", ""), 0);
}

Suite* words_suite(void) {
    Suite *s = suite_create("Words Test Suite");
    TCase *tc = tcase_create("Core");
    tcase_add_test(tc, test_reverse_string);
    tcase_add_test(tc, test_is_palindrome);
    tcase_add_test(tc, test_is_substring);
    tcase_add_test(tc, test_lcs);
    suite_add_tcase(s, tc);
    return s;
}